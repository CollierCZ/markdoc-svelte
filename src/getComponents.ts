import { Config } from "@markdoc/markdoc";
import FS from "fs";
import Path from "path";

export const getComponentImports = (
  schema: Config,
  componentDirPath: string,
): string => {
  let importStatement = "";

  const addImportsForSvelteComponents = (schemaArray: Config["tags"]) => {
    for (let name in schemaArray) {
      const tagOrNode = schemaArray[name];
      // Only include objects
      if (!tagOrNode || typeof tagOrNode === "string") continue;

      const renderName = tagOrNode["render"];
      /* Only when render string starts with a capital letter
         To represent Svelte components, not HTML tags and such
         Also ignores nodes with no render property */
      if (typeof renderName === "string" && /^\p{Lu}/u.test(renderName)) {
        // Check if the file exists and import if so
        const componentPath = Path.join(
          process.cwd(),
          `${componentDirPath}/${renderName}.svelte`,
        );
        if (FS.existsSync(componentPath)) {
          importStatement += `import ${renderName} from '${componentPath}';\n`;
        }
      }
    }
  };

  addImportsForSvelteComponents(schema.tags);
  addImportsForSvelteComponents(schema.nodes);

  return importStatement;
};
