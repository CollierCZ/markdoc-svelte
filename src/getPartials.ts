import MarkdocSource, { Config } from "@markdoc/markdoc";
import FS from "fs";
import Path from "path";

const getPartials = (
  partialsDirectory: string | undefined,
): Config["partials"] => {
  if (!partialsDirectory) return undefined;

  if (FS.existsSync(partialsDirectory)) {
    const files = FS.readdirSync(partialsDirectory);

    const partialsRecords = files.reduce(
      (recordObject, file) => {
        if (recordObject) {
          recordObject[file] = MarkdocSource.parse(
            FS.readFileSync(Path.join(partialsDirectory, file), "utf8"),
          );
        }
        return recordObject;
      },
      {} as Config["partials"],
    );

    return partialsRecords;
  } else {
    throw new Error(
      `Can't find the partials directory at '${partialsDirectory}'.`,
    );
  }
};

export default getPartials;
