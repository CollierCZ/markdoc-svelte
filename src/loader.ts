import { Config } from "@markdoc/markdoc";
import FS from "fs";
import Path from "path";

const DEFAULT_SCHEMA_PATH = "./markdoc";

// https://stackoverflow.com/questions/53799385/how-can-i-convert-a-windows-path-to-posix-path-using-node-path
const normalizeAbsolutePath = (absolutePath: string) => {
  return absolutePath.split(Path.sep).join(Path.posix.sep);
};

interface ConfigWithPartialsPath extends Omit<Config, "partials"> {
  partials?: string;
}

const loadSchema = async (
  schemaPath?: string
): Promise<ConfigWithPartialsPath> => {
  const schemaDirectory = Path.posix.resolve(schemaPath || DEFAULT_SCHEMA_PATH);

  const schemaDirectoryExists = FS.existsSync(schemaDirectory);

  let schemaCode = {};

  const getNormalizedPathInSchemaDirectory = (subDirectory: string) =>
    normalizeAbsolutePath(Path.posix.resolve(schemaDirectory, subDirectory));

  if (schemaDirectoryExists) {
    const readDirectory = async (directoryName: string) => {
      try {
        const module = getNormalizedPathInSchemaDirectory(directoryName);
        const moduleFile =
          FS.existsSync(`${module}.js`) || FS.existsSync(`${module}.ts`)
            ? `${module}`
            : `${module}/index`;
        const { default: schemaSection } = FS.existsSync(`${moduleFile}.js`)
          ? await import(`${moduleFile}.js`)
          : await import(`${moduleFile}.ts`);
        return schemaSection;
      } catch (error) {
        return {};
      }
    };

    const getPartialsDirectory = () => {
      try {
        const dirPath = getNormalizedPathInSchemaDirectory("partials");
        if (FS.existsSync(dirPath)) {
          return dirPath;
        }
        return "";
      } catch (error) {
        return "";
      }
    };

    schemaCode = {
      config: await readDirectory("config"),
      functions: await readDirectory("functions"),
      nodes: await readDirectory("nodes"),
      partials: getPartialsDirectory(),
      tags: await readDirectory("tags"),
      variables: await readDirectory("variables"),
    };
  } else if (
    // Throw an error only if the directory doesn't exist
    // AND a custom schema path is pased
    schemaPath &&
    schemaPath !== DEFAULT_SCHEMA_PATH
  ) {
    throw new Error(
      `Can't find the schema at '${schemaDirectory}' from the passed option 'schema: ${schemaPath}`
    );
  }

  return schemaCode;
};

export default loadSchema;
