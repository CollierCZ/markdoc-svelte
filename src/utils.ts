import MarkdownIt from "markdown-it";

const { escapeHtml } = MarkdownIt().utils;

const BRACKET_ESCAPE_TEST_RE = /[{}]/;
const BRACKET_ESCAPE_REPLACE_RE = /[{}]/g;
const BRACKET_REPLACEMENTS = {
  "{": "&lcub;",
  "}": "&rcub;",
};

const replaceMarkdocBrackets = (characters: string): string => {
  if (characters === "{" || characters === "}") {
    return BRACKET_REPLACEMENTS[characters];
  }
  return characters;
};

export const escapeMarkdocBrackets = (
  str: string,
  includeOtherHtml = true
): string => {
  if (BRACKET_ESCAPE_TEST_RE.test(str)) {
    return str.replace(BRACKET_ESCAPE_REPLACE_RE, replaceMarkdocBrackets);
  }
  return includeOtherHtml ? escapeHtml(str) : str;
};

const BRACKET_UNESCAPE_TEST_RE = /(&lcub;|&rcub;)/;
const BRACKET_UNESCAPE_REPLACE_RE = /(&lcub;|&rcub;)/g;

const replaceEscapedBracket = (escapedBracket: string) => {
  return (
    Object.keys(BRACKET_REPLACEMENTS).find(
      // @ts-ignore
      (key) => BRACKET_REPLACEMENTS[key] === escapedBracket
    ) || ""
  );
};

export const unescapeMarkdocBrackets = (str: string) => {
  if (BRACKET_UNESCAPE_TEST_RE.test(str)) {
    return str.replace(BRACKET_UNESCAPE_REPLACE_RE, replaceEscapedBracket);
  }
  return str;
};
