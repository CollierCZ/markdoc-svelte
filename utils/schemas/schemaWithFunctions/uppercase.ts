const uppercase = {
  transform(parameters: any) {
    const string = parameters[0];

    return typeof string === "string" ? string.toUpperCase() : string;
  },
};

export default uppercase;
