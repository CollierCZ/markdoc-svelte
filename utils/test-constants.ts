export const basicMarkdoc = `
# This is some basic Markdoc

With a paragraph.

Some text **in bold** and *in italic*.

And a [link](https://example.com).

## More fancy stuff

And even a table with a nested list:

{% table %}
* Table header 1
* Table header 2
---
* * Row 1 Cell 1 Item 1
  * Row 1 Cell 1 Item 2
* Row 1 Cell 2
---
* Row 2 Cell 1
* Row 2 cell 2
{% /table %}
`;

export const markdocWithFrontmatter = `---
title: A test
description: A longer idea
keywords:
  - one
  - two
---

Content. Such great content.
`;

export const markdocWithComponent = `${basicMarkdoc}
{% button text="Hello world" /%}
`