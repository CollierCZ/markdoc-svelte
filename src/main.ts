import MarkdocSource, { Config, validate } from "@markdoc/markdoc";
import yaml from "js-yaml";
import render from "./render";
import loadSchema from "./loader";
import { getComponentImports } from "./getComponents";
import getPartials from "./getPartials";

interface Options {
  extensions?: string[];
  layout?: string;
  validationLevel?: "debug" | "info" | "warning" | "error" | "critical";
  schema?: string;
  functions?: Config["functions"];
  nodes?: Config["nodes"];
  partials?: string;
  tags?: Config["tags"];
  variables?: Config["variables"];
}

interface PreprocessorReturn {
  code: string;
  data?: Record<string, unknown>;
  map?: string;
}

interface Preprocessor {
  markup: (args: {
    content: string;
    filename: string;
  }) => Promise<PreprocessorReturn | undefined>;
}

/**
 * A Svelte preprocessor for Markdoc files
 */
export const markdoc = (options: Options = {}): Preprocessor => {
  const layoutPath = options.layout;
  const schemaPath = options.schema;
  const extensions = options.extensions || [".markdown", ".md"];
  const validationLevel = options.validationLevel || "error";
  const {
    functions,
    nodes,
    partials: partialsDirectory,
    tags,
    variables,
  } = options;

  return {
    markup: async ({ content = "", filename = "" }) => {
      if (!extensions.find((extension) => filename.endsWith(extension))) return;

      const ast = MarkdocSource.parse(content);

      const isFrontmatter = Boolean(ast.attributes.frontmatter);
      const frontmatter = isFrontmatter
        ? (yaml.load(ast.attributes.frontmatter) as Record<string, unknown>)
        : {};

      const schemaFromPath = await loadSchema(schemaPath);

      const {
        partials: partialsDirectoryFromSchema,
        variables: variablesFromSchema,
        ...schemaFromPathWithoutPartials
      } = schemaFromPath;

      // Include schema parts passed as options
      // But ignore if undefined
      // Leave out partials until directory processed
      const schemaWithoutPartials = {
        ...schemaFromPathWithoutPartials,
        ...(functions && { functions }),
        ...(nodes && { nodes }),
        ...(tags && { tags }),
      };

      const markdocConfig = {
        ...schemaWithoutPartials,
        variables: { frontmatter, ...(variables || variablesFromSchema) },
        partials:
          partialsDirectory || schemaFromPath["partials"]
            ? getPartials(partialsDirectory || schemaFromPath["partials"])
            : undefined,
      };

      // Check if Markdoc is valid
      const errorLevelsMap = new Map<Options["validationLevel"], number>([
        ["debug", 0],
        ["info", 1],
        ["warning", 2],
        ["error", 3],
        ["critical", 4],
      ]);
      const errors = MarkdocSource.validate(ast, markdocConfig);
      const breakingLevel = errorLevelsMap.get(validationLevel)!;
      const areErrorsAtBreakingLevel = errors.find(
        (error) => errorLevelsMap.get(error.error.level)! >= breakingLevel
      );
      if (areErrorsAtBreakingLevel) {
        throw new Error(
          `The file at ${filename} is invalid with ${errors.length} error${errors.length > 1 ? "s" : ""}:\n- ${errors.map((error) => error.error.message).join("\n- ")}`
        );
      }

      const transformedContent = await MarkdocSource.transform(
        ast,
        markdocConfig
      );

      const svelteContent = render(transformedContent);
      const frontmatterString = isFrontmatter
        ? `<script context="module">\n` +
          `\texport const metadata = ${JSON.stringify(frontmatter)};\n` +
          `\tconst { ${Object.keys(frontmatter as Record<string, unknown>).join(
            ", "
          )} } = metadata;\n` +
          "</script>\n"
        : "";

      const componentsString = getComponentImports(
        schemaWithoutPartials,
        "/src/lib/components"
      );
      const layoutOpenString =
        layoutPath || componentsString
          ? `
<script>
  ${layoutPath ? `import Layout_DEFAULT from '${layoutPath}';` : ""}
  ${componentsString}
</script>
${
  layoutPath ? `<Layout_DEFAULT${isFrontmatter ? ` {...metadata}` : ""}>\n` : ""
}`
          : "";

      const layoutCloseString = layoutPath ? "</Layout_DEFAULT>\n" : "";

      const code =
        frontmatterString +
        layoutOpenString +
        svelteContent +
        layoutCloseString;

      return {
        code: code,
        data: frontmatter,
      };
    },
  };
};
