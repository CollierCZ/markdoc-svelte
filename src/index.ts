import { markdoc } from "./main";
import { escapeMarkdocBrackets, unescapeMarkdocBrackets } from "./utils";

export { escapeMarkdocBrackets, markdoc, unescapeMarkdocBrackets };
