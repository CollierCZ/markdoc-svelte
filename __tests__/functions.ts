import { markdoc } from "../src/main";
import { markdocWithFunction } from "../utils/test-constants";
import uppercase from "../utils/schemas/schemaWithFunctions/uppercase"

test(`processes variables passed with schema path`, async () => {
  console.log("path")
  expect(
    await markdoc({ schema: "./utils/schemas/schemaWithFunctions" }).markup({
      content: markdocWithFunction,
      filename: "test.md",
    }),
  ).toMatchSnapshot();
});

test(`processes variables passed directly`, async () => {
  console.log("direct")
  expect(
    await markdoc({
      functions: { uppercase },
    }).markup({
      content: markdocWithFunction,
      filename: "test.md",
    }),
  ).toMatchSnapshot();
});
