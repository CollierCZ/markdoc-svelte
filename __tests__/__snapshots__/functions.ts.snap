// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processes variables passed directly 1`] = `
{
  "code": "<article><p>This is some text</p><p>This text will be transformed: UPPERCASE</p></article>",
  "data": {},
}
`;

exports[`processes variables passed with schema path 1`] = `
{
  "code": "<article><p>This is some text</p><p>This text will be transformed: UPPERCASE</p></article>",
  "data": {},
}
`;
